packages/event-bus-redis/src/services/event-bus-redis.ts

```diff
@@ -0,0 +1,228 @@
+import {
+  AbstractEventBusModuleService,
+  ConfigModule,
+  ConfigurableModuleDeclaration,
+  Logger,
+  MODULE_RESOURCE_TYPE
+} from "@medusajs/medusa"
+import { Job, Queue, Worker } from "bullmq"
+import Redis from "ioredis"
+import { isDefined, MedusaError } from "medusa-core-utils"
+
+type InjectedDependencies = {
+  logger: Logger
+  configModule: ConfigModule
+}
+
+type EventBusRedisModuleOptions = {
+  queueName?: string
+  workerName?: string
+  queuePrefix?: string
+  workerPrefix?: string
+  redisUrl?: string
+}
+
+type BullJob<T> = {
+  data: {
+    eventName: string
+    data: T
+    completedSubscriberIds: string[] | undefined
+  }
+} & Job
+
+type EmitOptions = {
+  delay?: number
+  attempts: number
+  backoff?: {
+    type: "fixed" | "exponential"
+    delay: number
+  }
+}
+
+/**
+ * Can keep track of multiple subscribers to different events and run the
+ * subscribers when events happen. Events will run asynchronously.
+ */
+export default class RedisEventBusService extends AbstractEventBusModuleService {
+  protected readonly config_: ConfigModule
+  protected readonly moduleOptions_: EventBusRedisModuleOptions
+  protected readonly moduleDeclaration_: ConfigurableModuleDeclaration
+  protected readonly logger_: Logger
+
+  protected queue_: Queue
+  protected shouldEnqueuerRun: boolean
+  protected enqueue_: Promise<void>
+
+  constructor(
+    { logger, configModule }: InjectedDependencies,
+    moduleOptions: EventBusRedisModuleOptions = {},
+    moduleDeclaration: ConfigurableModuleDeclaration,
+    singleton = true
+  ) {
+    // @ts-ignore
+    super(...arguments)
+
+    if (moduleDeclaration?.resources !== MODULE_RESOURCE_TYPE.SHARED) {
+      throw new MedusaError(
+        MedusaError.Types.INVALID_ARGUMENT,
+        "At the moment this module can only be used with shared resources"
+      )
+    }
+
+    this.moduleOptions_ = moduleOptions
+    this.moduleDeclaration_ = moduleDeclaration
+    this.config_ = configModule
+    this.logger_ = logger
+
+    if (singleton && moduleOptions.redisUrl) {
```

Comment:
thought: guess this guard can be removed altogether since there is no singleton and we have already verified that `redisUrl` exists